src/main/java/org/opensearch/agent/tools/KnowledgeBaseTool.java

```diff
@@ -0,0 +1,163 @@
+import org.opensearch.ml.common.input.MLInput;
+import org.opensearch.ml.common.output.model.ModelTensor;
+import org.opensearch.ml.common.output.model.ModelTensorOutput;
+import org.opensearch.ml.common.output.model.ModelTensors;
+import org.opensearch.ml.common.spi.tools.Tool;
+import org.opensearch.ml.common.spi.tools.ToolAnnotation;
+import org.opensearch.ml.common.transport.prediction.MLPredictionTaskAction;
+import org.opensearch.ml.common.transport.prediction.MLPredictionTaskRequest;
+
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+@Setter
+@Getter
+@ToolAnnotation(KnowledgeBaseTool.TYPE)
+public class KnowledgeBaseTool implements Tool {
+
+    public static final String TYPE = "KnowledgeBaseTool";
+
+    private String name = TYPE;
+
+    private String description = DEFAULT_DESCRIPTION;
+
+    private String modelId;
+
+    public static final String INPUT_FIELD = "text";
+
+    private Client client;
+
+    private static final String DEFAULT_DESCRIPTION = "Use this tool to retrieve related documentation on OpenSearch.";
+
+    public KnowledgeBaseTool(Client client, String modelId) {
+        this.client = client;
+        this.modelId = modelId;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> void run(Map<String, String> parameters, ActionListener<T> listener) {
+        String text = parameters.get(INPUT_FIELD);
+        if (StringUtils.isBlank(text)) {
+            throw new IllegalArgumentException("[" + INPUT_FIELD + "] is null or empty, can not process it in knowledge base tool.");
+        }
+
+        RemoteInferenceInputDataSet inputDataSet = RemoteInferenceInputDataSet
+            .builder()
+            .parameters(Collections.singletonMap(INPUT_FIELD, text))
+            .build();
+        MLInput mlInput = MLInput.builder().algorithm(FunctionName.REMOTE).inputDataset(inputDataSet).build();
+        ActionRequest request = new MLPredictionTaskRequest(modelId, mlInput, null);
+
+        client.execute(MLPredictionTaskAction.INSTANCE, request, ActionListener.wrap(mlTaskResponse -> {
+            ModelTensorOutput modelTensorOutput = (ModelTensorOutput) mlTaskResponse.getOutput();
+            ModelTensors modelTensors = modelTensorOutput.getMlModelOutputs().get(0);
+            ModelTensor modelTensor = modelTensors.getMlModelTensors().get(0);
+            Map<String, Object> dataAsMap = (Map<String, Object>) modelTensor.getDataAsMap();
+            if (dataAsMap.get("retrievalResults") == null) {
+                listener.onFailure(new IllegalStateException("Remote endpoint fails to retrieve documents"));
+                return;
+            }
+            ArrayList<Map<String, Object>> retrievedDocuments = (ArrayList<Map<String, Object>>) dataAsMap.get("retrievalResults");
+            StringBuilder formattedDocuments = new StringBuilder();
+            for (Map<String, Object> retrievedDocument : retrievedDocuments) {
```

Comment:
If knowledge base endpoint returns an error like exceed limitation (I don't know, we may need a survey), we will return empty result, does this meet expectation?

Reply:
Any idea to mock an error exception from bedrock knowledge base?

Reply:
> Any idea to mock an error exception from bedrock knowledge base?

We can make a easy real environment exception by remove the model access to the embedding model and then call the API.
Maybe return the origin exception information is enough?

Reply:
Good idea. I can start testing once we have plan to merge this PR. I will mark this PR as draft.